### PR TITLE
Added support for Pragmas

### LINKED ISSUE #73: Add support for pragmas

### PR DESCRIPTION
**Pull request by fferreira**
Fixes #73

### CODE REVIEW

**Comment by fangyi-zhou** [comment]
Can you rebase?

File: bin/main.ml [hunk 1]
```diff
@@ -2,6 +2,13 @@ open Base
 open Stdio
 open Nuscrlib
 
+let version_string () = "%%VERSION%%"
+
+let usage () =
+  "usage: "
+  ^ (Sys.get_argv ()).(0)
+  ^ " [-enum][-verbose][-fsm Role@Protocol][-project Role@Protocol] file"
```
Comment (fangyi-zhou) [inline bin/main.ml]: We should probably add other usages in another commit

File: lib/parser.mly [hunk 2]
```diff
@@ -40,7 +40,8 @@
 %token DO_KW
 
 (* ---------------------------------------- *)
-%start <Syntax.scr_module> scr_module
+%start <Syntax.scr_module> doc
+%start <(string * string option) list> pragmas
```
Comment (fangyi-zhou) [inline lib/parser.mly]: Syntax.pragma

File: CHANGES.md [hunk 3]
```diff
@@ -2,6 +2,7 @@
```
Comment (fangyi-zhou) [inline CHANGES.md]: Did you push the commit replaceing the Unreleased to 1.1.0? This changelog entry should go to the new "Unreleased" and the current "Unreleased" should be "1.1.0"

File: bin/main.ml [hunk 4]
```diff
@@ -58,8 +65,19 @@ let gen_output ast f = function
       print_endline res
   | _ -> ()
 
+let process_pragmas (pragmas : Syntax.pragmas) : unit =
+  let process_global_pragma (k, v) =
+    match (k, v) with
+    | "PrintUsage", _ -> usage () |> print_endline
+    | "ShowPragmas", _ -> Syntax.show_pragmas pragmas |> print_endline
+    | _ -> ()
```
Comment (fangyi-zhou) [inline bin/main.ml]: Do we simply ignore other pragmas for now?

File: bin/main.ml [hunk 5]
```diff
@@ -58,8 +65,19 @@ let gen_output ast f = function
       print_endline res
   | _ -> ()
 
+let process_pragmas (pragmas : Syntax.pragmas) : unit =
+  let process_global_pragma (k, v) =
+    match (k, v) with
+    | "PrintUsage", _ -> usage () |> print_endline
```
Comment (fangyi-zhou) [inline bin/main.ml]: Is PrintUsage particular? Or you just need a random example?
It seems rather unusual to require printing usage via a pragma

Comment (fferreira) [reply bin/main.ml]: it is a random example

File: bin/main.ml (hunk 1 above)
Comment (fferreira) [reply bin/main.ml]: this is another issue

File: CHANGES.md [hunk 6]
```diff
@@ -17,6 +17,7 @@
 
 ## Added
 - Export code generation APIs in Lib
+- Added support for pragmas in code
```
Comment (fangyi-zhou) [inline CHANGES.md]: This should be moved up

File: lib/err.ml [hunk 7]
```diff
@@ -25,6 +26,7 @@ exception UserError of user_error
 (** UserError is a user error and should be reported back so it can be fixed *)
 
 let show_user_error = function
+  | UnknownPragma prg -> "Unknown pragma:" ^ prg
```
Comment (fangyi-zhou) [inline lib/err.ml]:
```suggestion
  | UnknownPragma prg -> "Unknown pragma: " ^ prg
```

File: CHANGES.md (hunk 6 above)
Comment (fferreira) [reply CHANGES.md]: done

**Pull request merged**